crml/cennzx-spot/src/lib.rs

```diff
@@ -436,58 +459,51 @@ impl<T: Trait> Module<T> {
+	/// `asset_to_payout` is the asset to be paid out in exchange for the sale of `asset_to_sell` (the final sale value is given in this asset)
+	pub fn get_sell_price(
+		asset_to_sell: T::AssetId,
+		amount_to_sell: T::Balance,
+		asset_to_payout: T::AssetId,
+	) -> Result<T::Balance, DispatchError> {
+		ensure!(asset_to_sell != asset_to_payout, Error::<T>::AssetCannotSwapForItself);
 
-		let input_reserve_hp = HighPrecisionUnsigned::from(T::BalanceToUnsignedInt::from(input_reserve).into());
-		let output_reserve_hp = HighPrecisionUnsigned::from(T::BalanceToUnsignedInt::from(output_reserve).into());
-		let input_amount_scaled_hp = HighPrecisionUnsigned::from(LowPrecisionUnsigned::from(input_amount_scaled));
-		let denominator_hp = input_amount_scaled_hp + input_reserve_hp;
-		let price_hp = output_reserve_hp
-			.saturating_mul(input_amount_scaled_hp)
-			.checked_div(denominator_hp)
-			.ok_or::<Error<T>>(Error::<T>::DivideByZero)?;
+		// Find the value of `amount_to_sell` of `asset_to_sell` in terms of core asset
+		// (how much core asset is the sale worth?)
+		let core_asset_amount = if asset_to_sell == Self::core_asset_id() {
+			amount_to_sell
+		} else {
+			Self::get_asset_to_core_sell_price(&asset_to_sell, amount_to_sell)?
+		};
 
-		let price_lp_result: Result<LowPrecisionUnsigned, &'static str> = LowPrecisionUnsigned::try_from(price_hp);
-		ensure!(price_lp_result.is_ok(), Error::<T>::Overflow);
-		let price_lp = price_lp_result.unwrap();
+		// Skip payout asset price if asset to be paid out is core
+		// (how much `asset_to_payout` is the sale worth?)
+		let payout_asset_value = if asset_to_payout == Self::core_asset_id() {
+			core_asset_amount
+		} else {
+			Self::get_core_to_asset_sell_price(&asset_to_payout, core_asset_amount)?
+		};
 
-		let price = T::UnsignedIntToBalance::from(price_lp).into();
-		ensure!(output_reserve > price, Error::<T>::InsufficientAssetReserve);
-		Ok(price)
+		Ok(payout_asset_value)
 	}
 
 	/// `asset_id` - Trade asset
-	/// `buy_amount` - Amount of output core
-	/// Returns the amount of trade assets needed to buy `buy_amount` core assets.
-	pub fn get_asset_to_core_buy_price(
+	/// `amount_sold` - Amount of the trade asset to sell
+	/// Returns amount of core that can be bought with input assets.
+	pub fn get_asset_to_core_sell_price(
 		asset_id: &T::AssetId,
-		buy_amount: T::Balance,
+		sell_amount: T::Balance,
 	) -> sp_std::result::Result<T::Balance, DispatchError> {
-		ensure!(buy_amount > Zero::zero(), Error::<T>::BuyAmountNotPositive);
-
-		let core_asset_id = Self::core_asset_id();
-		let exchange_address = T::ExchangeAddressGenerator::exchange_address_for(*asset_id);
-
-		let core_asset_reserve = <pallet_generic_asset::Module<T>>::free_balance(&core_asset_id, &exchange_address);
-		let trade_asset_reserve = <pallet_generic_asset::Module<T>>::free_balance(&asset_id, &exchange_address);
+		ensure!(
+			sell_amount > Zero::zero(),
+			Error::<T>::AssetToCoreSellAmountNotAboveZero
```

Comment:
Yeah we should squash all of these error types into a `CannotTradeZero`

Reply:
Will get to it soon - I have a couple more PRs in mind for this module 😉